File: kerb_chain/engagement.py
"""
Engagement state — the in-memory graph kerb-chain walks.

An Engagement starts with whatever kerb-map already discovered
(findings, domain info, the operator's bind credential) and grows as
plays run: each play can deposit credentials, hashes, Kerberos
tickets, owned hosts, and certificates. Subsequent plays consult the
state to decide what to run next, and the placeholder substitution in
play command templates pulls from here too.

The state is intentionally a plain Python dataclass tree rather than a
graph database — kerb-chain's MVP target is a single-engagement
operator session, not a multi-team C2. If we later need
cross-engagement queries (or a UI on top), pivoting to Neo4j /
memgraph is one adapter.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path

# ────────────────────────────────────────────────────────────────────── #
#  Loot                                                                  #
# ────────────────────────────────────────────────────────────────────── #


@dataclass
class Credential:
    """A username + secret + the play that produced it. Either ``password``
    or ``nt_hash`` is set; never both blank."""

    username:  str
    domain:    str
    password:  str | None = None
    nt_hash:   str | None = None
    source:    str = "operator"          # play name, or 'operator' for the seed cred
    obtained_at: str = ""

    def __post_init__(self):
        if not self.obtained_at:
            self.obtained_at = datetime.now(timezone.utc).isoformat(timespec="seconds")
        if not (self.password or self.nt_hash):
            raise ValueError("Credential needs at least one of password / nt_hash")

    @property
    def upn(self) -> str:
        return f"{self.username}@{self.domain}"


@dataclass
class Ticket:
    """A Kerberos TGT or service ticket on disk (ccache or .kirbi)."""

    principal: str
    path:      Path
    kind:      str = "tgt"               # 'tgt' or 'service'
    spn:       str | None = None         # for service tickets
    source:    str = "operator"
    obtained_at: str = ""

    def __post_init__(self):
        if not self.obtained_at:
            self.obtained_at = datetime.now(timezone.utc).isoformat(timespec="seconds")


@dataclass
class Certificate:
    """A captured PFX + (optional) password — typically from ESC1/4/8 or
    Shadow Credentials chains."""

    subject:   str
    pfx_path:  Path
    pfx_pass:  str | None = None
    source:    str = ""
    obtained_at: str = ""

    def __post_init__(self):
        if not self.obtained_at:
            self.obtained_at = datetime.now(timezone.utc).isoformat(timespec="seconds")


@dataclass
class OwnedHost:
    """A host kerb-chain has popped a session on (or already had access
    to). 'session' is whatever the play produced — ssh, smbexec, wmiexec
    output handle, etc."""

    name:       str
    ip:         str | None = None
    via_play:   str = ""
    obtained_at: str = ""
    notes:      str = ""

    def __post_init__(self):
        if not self.obtained_at:
            self.obtained_at = datetime.now(timezone.utc).isoformat(timespec="seconds")


@dataclass
class Loot:
    """All collected loot in one place. Plays read and append; the
    Engagement passes a reference to each play's command-template
    renderer, so any of these can be referenced as ``{{loot.foo}}``."""

    credentials: list[Credential]   = field(default_factory=list)
    tickets:     list[Ticket]       = field(default_factory=list)
    certificates: list[Certificate] = field(default_factory=list)
    owned_hosts: list[OwnedHost]    = field(default_factory=list)
    files:       dict[str, Path]    = field(default_factory=dict)

    def has_creds_for(self, username: str) -> bool:
        return any(c.username.lower() == username.lower() for c in self.credentials)

    def best_cred(self, username: str | None = None) -> Credential | None:
        """The newest credential, optionally filtered by username.
        Used by the default placeholder ``{{operator}}`` in playbooks."""
        candidates = self.credentials
        if username:
            candidates = [c for c in candidates if c.username.lower() == username.lower()]
        return candidates[-1] if candidates else None


# ────────────────────────────────────────────────────────────────────── #
#  Engagement                                                            #
# ────────────────────────────────────────────────────────────────────── #


@dataclass
class PlayRecord:
    """One play execution — what ran, when, what came out."""

    play:       str
    command:    list[str]
    started_at: str
    finished_at: str = ""
    exit_code:  int | None = None
    stdout:     str = ""
    stderr:     str = ""
    loot_added: dict[str, int] = field(default_factory=dict)
    skipped:    str | None = None        # reason if not executed


@dataclass
class Engagement:
    """The full state kerb-chain operates on. Construct with
    ``Engagement.from_findings(...)`` rather than the bare constructor;
    that ensures sensible defaults for run_dir, operator cred, dry-run."""

    findings:   list[dict]
    domain:     str
    dc_ip:      str
    base_dn:    str
    domain_sid: str | None
    operator:   Credential | None       # the bind credential kerb-map used
    loot:       Loot                  = field(default_factory=Loot)
    history:    list[PlayRecord]      = field(default_factory=list)
    run_dir:    Path                  = field(default_factory=Path.cwd)
    dry_run:    bool                  = False

    @classmethod
    def from_findings(
        cls,
        findings: list[dict],
        *,
        domain:    str = "",
        dc_ip:     str = "",
        base_dn:   str = "",
        domain_sid: str | None = None,
        operator_cred: Credential | None = None,
        run_dir:   str | Path | None = None,
        dry_run:   bool = False,
    ) -> Engagement:
        # When the findings carry a meta block (the kerb-map JSON shape),
        # use it as the source of truth for domain/dc_ip — otherwise the
        # operator must pass them explicitly.
        meta = {}
        if findings and isinstance(findings, list) and findings[0].get("__meta__"):
            meta = findings[0]["__meta__"]

        domain  = domain or meta.get("domain", "")
        dc_ip   = dc_ip or meta.get("dc_ip", "")
        base_dn = base_dn or meta.get("base_dn", "")
        domain_sid = domain_sid or meta.get("domain_sid")

        # Pull domain_sid out of the first finding that has one, if
        # we still don't know it — DCSync / Shadow Creds findings carry it.
        if not domain_sid:
            for f in findings:
                ds = (f.get("data") or {}).get("domain_sid")
                if ds:
                    domain_sid = ds
                    break

        out_dir = Path(run_dir or _default_run_dir(domain))
        out_dir.mkdir(parents=True, exist_ok=True)

        loot = Loot()
        if operator_cred:
            loot.credentials.append(operator_cred)

        return cls(
            findings=findings,
            domain=domain,
            dc_ip=dc_ip,
            base_dn=base_dn,
            domain_sid=domain_sid,
            operator=operator_cred,
            loot=loot,
            run_dir=out_dir,
            dry_run=dry_run,
        )

    # ------------------------------------------------------------------ #
    #  Placeholder substitution                                          #
    # ------------------------------------------------------------------ #

    def render(self, template: str, *, finding: dict | None = None) -> str:
        """Substitute ``{{domain}}``, ``{{dc_ip}}``, ``{{operator_user}}``,
        ``{{operator_pass}}``, ``{{run_dir}}``, ``{{finding.target}}``,
        ``{{finding.data.X}}``, etc. into ``template``.

        Deliberately a tiny custom resolver rather than Jinja — playbooks
        should be auditable and a sandbox-friendly subset is enough.
        Unknown placeholders are left as-is so the operator notices.
        """
        ctx = self._render_context(finding)
        out = template
        # Two passes so that {{a}} can resolve to "{{b}}" once and we still
        # try {{b}} the second time. Cap at a few iterations to stop cycles.
        for _ in range(4):
            new = out
            for key, value in ctx.items():
                new = new.replace("{{" + key + "}}", str(value))
            if new == out:
                break
            out = new
        return out

    def _render_context(self, finding: dict | None) -> dict[str, object]:
        op = self.loot.best_cred() or self.operator
        ctx: dict[str, object] = {
            "domain":         self.domain,
            "dc_ip":          self.dc_ip,
            "base_dn":        self.base_dn,
            "domain_sid":     self.domain_sid or "",
            "run_dir":        str(self.run_dir),
            "operator_user":  op.username if op else "",
            "operator_pass":  op.password if op else "",
            "operator_hash":  op.nt_hash if op else "",
        }
        if finding:
            ctx["finding.target"]   = finding.get("target", "")
            ctx["finding.attack"]   = finding.get("attack", "")
            ctx["finding.severity"] = finding.get("severity", "")
            for k, v in (finding.get("data") or {}).items():
                ctx[f"finding.data.{k}"] = v
        return ctx

    # ------------------------------------------------------------------ #
    #  Persistence                                                       #
    # ------------------------------------------------------------------ #

    def write_journal(self) -> Path:
        """Write the run history + loot to ``run_dir/journal.json``.
        Caller decides when (typically end-of-run, but per-play is fine
        for long sessions where you want progress on disk)."""
        out = self.run_dir / "journal.json"
        payload = {
            "domain":     self.domain,
            "dc_ip":      self.dc_ip,
            "domain_sid": self.domain_sid,
            "loot": {
                "credentials":  [_dataclass_dict(c) for c in self.loot.credentials],
                "tickets":      [_dataclass_dict(t) for t in self.loot.tickets],
                "certificates": [_dataclass_dict(c) for c in self.loot.certificates],
                "owned_hosts":  [_dataclass_dict(h) for h in self.loot.owned_hosts],
                "files":        {k: str(v) for k, v in self.loot.files.items()},
            },
            "history": [_dataclass_dict(p) for p in self.history],
        }
        out.write_text(json.dumps(payload, indent=2, default=str))
        return out


def _default_run_dir(domain: str) -> Path:
    ts = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    domain_part = domain.replace(".", "_") if domain else "engagement"
    base = Path(os.environ.get("KERB_CHAIN_HOME") or (Path.home() / ".kerb-chain"))
    return base / "runs" / f"{domain_part}_{ts}"


def _dataclass_dict(obj) -> dict:
    """Best-effort dataclass → dict that handles Path and the dataclasses
    we actually use. Avoids importing dataclasses.asdict (which complains
    about non-dataclass fields the operator might add)."""
    out: dict[str, object] = {}
    for k, v in vars(obj).items():
        if isinstance(v, Path):
            v = str(v)
        out[k] = v
    return out

File: kerb-chain-rs/src/engagement.rs
//! Engagement state — the in-memory graph kerb-chain walks.
//!
//! Direct port of `kerb_chain/engagement.py`. The serialised
//! `journal.json` format is byte-compatible with the Python version
//! (verified by integration test), so operators can run either runtime
//! against the same engagement directory and the history interleaves.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use anyhow::{Result, anyhow};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::findings::Finding;


// ─────────────────────────────────────────────────────────────────── //
//  Loot                                                               //
// ─────────────────────────────────────────────────────────────────── //


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Credential {
    pub username: String,
    pub domain:   String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nt_hash:  Option<String>,
    #[serde(default)]
    pub source:   String,
    #[serde(default)]
    pub obtained_at: String,
}

impl Credential {
    pub fn new(username: impl Into<String>, domain: impl Into<String>) -> Self {
        Self {
            username:    username.into(),
            domain:      domain.into(),
            password:    None,
            nt_hash:     None,
            source:      "operator".to_string(),
            obtained_at: now_iso(),
        }
    }

    pub fn with_password(mut self, pw: impl Into<String>) -> Self {
        self.password = Some(pw.into());
        self
    }

    pub fn with_hash(mut self, nt_hash: impl Into<String>) -> Self {
        self.nt_hash = Some(nt_hash.into());
        self
    }

    pub fn with_source(mut self, source: impl Into<String>) -> Self {
        self.source = source.into();
        self
    }

    pub fn validate(&self) -> Result<()> {
        if self.password.is_none() && self.nt_hash.is_none() {
            return Err(anyhow!("Credential needs at least one of password / nt_hash"));
        }
        Ok(())
    }

    pub fn upn(&self) -> String {
        format!("{}@{}", self.username, self.domain)
    }
}


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Ticket {
    pub principal:   String,
    pub path:        PathBuf,
    pub kind:        String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub spn:         Option<String>,
    #[serde(default)]
    pub source:      String,
    #[serde(default)]
    pub obtained_at: String,
}


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Certificate {
    pub subject:   String,
    pub pfx_path:  PathBuf,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pfx_pass:  Option<String>,
    #[serde(default)]
    pub source:    String,
    #[serde(default)]
    pub obtained_at: String,
}


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OwnedHost {
    pub name:        String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ip:          Option<String>,
    #[serde(default)]
    pub via_play:    String,
    #[serde(default)]
    pub obtained_at: String,
    #[serde(default)]
    pub notes:       String,
}


#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Loot {
    #[serde(default)]
    pub credentials:  Vec<Credential>,
    #[serde(default)]
    pub tickets:      Vec<Ticket>,
    #[serde(default)]
    pub certificates: Vec<Certificate>,
    #[serde(default)]
    pub owned_hosts:  Vec<OwnedHost>,
    #[serde(default)]
    pub files:        BTreeMap<String, PathBuf>,
}

impl Loot {
    pub fn has_creds_for(&self, username: &str) -> bool {
        let lc = username.to_ascii_lowercase();
        self.credentials.iter().any(|c| c.username.to_ascii_lowercase() == lc)
    }

    /// Newest credential, optionally filtered by username. Used by the
    /// default `{{operator_user}}` / `{{operator_pass}}` placeholders.
    pub fn best_cred(&self, username: Option<&str>) -> Option<&Credential> {
        self.credentials
            .iter()
            .rev()
            .find(|c| match username {
                None    => true,
                Some(u) => c.username.eq_ignore_ascii_case(u),
            })
    }
}


// ─────────────────────────────────────────────────────────────────── //
//  Engagement + history                                               //
// ─────────────────────────────────────────────────────────────────── //


#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PlayRecord {
    pub play:        String,
    #[serde(default)]
    pub command:     Vec<String>,
    pub started_at:  String,
    #[serde(default)]
    pub finished_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exit_code:   Option<i32>,
    #[serde(default)]
    pub stdout:      String,
    #[serde(default)]
    pub stderr:      String,
    #[serde(default)]
    pub loot_added:  BTreeMap<String, usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skipped:     Option<String>,
}


#[derive(Debug)]
pub struct Engagement {
    pub findings:   Vec<Finding>,
    pub domain:     String,
    pub dc_ip:      String,
    pub base_dn:    String,
    pub domain_sid: Option<String>,
    pub operator:   Option<Credential>,
    pub loot:       Loot,
    pub history:    Vec<PlayRecord>,
    pub run_dir:    PathBuf,
    pub dry_run:    bool,
}

#[derive(Debug, Default)]
pub struct EngagementOpts {
    pub domain:     Option<String>,
    pub dc_ip:      Option<String>,
    pub base_dn:    Option<String>,
    pub domain_sid: Option<String>,
    pub operator:   Option<Credential>,
    pub run_dir:    Option<PathBuf>,
    pub dry_run:    bool,
}

impl Engagement {
    /// Build an engagement from a findings list. Mirrors
    /// `Engagement.from_findings(...)` on the Python side: pulls
    /// `domain` / `dc_ip` from the first finding's `__meta__` block
    /// when not explicitly provided, derives `domain_sid` from any
    /// finding that carries it, materialises `run_dir` on disk, and
    /// seeds the loot with the operator credential.
    pub fn from_findings(findings: Vec<Finding>, opts: EngagementOpts) -> Result<Self> {
        let meta = findings
            .first()
            .and_then(|f| f.get("__meta__"))
            .and_then(|v| v.as_object())
            .cloned()
            .unwrap_or_default();

        let pick_str = |opt: Option<String>, key: &str| -> String {
            opt.unwrap_or_else(|| {
                meta.get(key)
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string())
                    .unwrap_or_default()
            })
        };

        let domain  = pick_str(opts.domain, "domain");
        let dc_ip   = pick_str(opts.dc_ip,  "dc_ip");
        let base_dn = pick_str(opts.base_dn, "base_dn");

        // domain_sid: prefer explicit > meta > scan through findings.data
        let mut domain_sid = opts.domain_sid.or_else(|| {
            meta.get("domain_sid").and_then(|v| v.as_str()).map(|s| s.to_string())
        });
        if domain_sid.is_none() {
            for f in &findings {
                if let Some(ds) = f
                    .get("data")
                    .and_then(|v| v.as_object())
                    .and_then(|d| d.get("domain_sid"))
                    .and_then(|v| v.as_str())
                {
                    domain_sid = Some(ds.to_string());
                    break;
                }
            }
        }

        let run_dir = opts.run_dir.unwrap_or_else(|| default_run_dir(&domain));
        std::fs::create_dir_all(&run_dir)?;

        let mut loot = Loot::default();
        if let Some(cred) = opts.operator.clone() {
            cred.validate()?;
            loot.credentials.push(cred);
        }

        Ok(Self {
            findings,
            domain,
            dc_ip,
            base_dn,
            domain_sid,
            operator: opts.operator,
            loot,
            history: Vec::new(),
            run_dir,
            dry_run: opts.dry_run,
        })
    }

    /// Substitute `{{...}}` placeholders into a template.
    /// Resolves the engagement context plus per-finding fields
    /// (`finding.target`, `finding.attack`, `finding.severity`,
    /// `finding.data.*`). Unknown placeholders are left in place so
    /// operators notice typos rather than silently getting empty strings.
    pub fn render(&self, template: &str, finding: Option<&Finding>) -> String {
        let ctx = self.render_context(finding);
        let mut out = template.to_string();
        for _ in 0..4 {
            let prev = out.clone();
            for (key, value) in &ctx {
                let needle = format!("{{{{{key}}}}}");
                if out.contains(&needle) {
                    out = out.replace(&needle, value);
                }
            }
            if out == prev {
                break;
            }
        }
        out
    }

    fn render_context(&self, finding: Option<&Finding>) -> BTreeMap<String, String> {
        let mut ctx: BTreeMap<String, String> = BTreeMap::new();
        ctx.insert("domain".into(),     self.domain.clone());
        ctx.insert("dc_ip".into(),      self.dc_ip.clone());
        ctx.insert("base_dn".into(),    self.base_dn.clone());
        ctx.insert("domain_sid".into(), self.domain_sid.clone().unwrap_or_default());
        ctx.insert("run_dir".into(),    self.run_dir.display().to_string());

        let op = self.loot.best_cred(None).or(self.operator.as_ref());
        ctx.insert("operator_user".into(), op.map(|c| c.username.clone()).unwrap_or_default());
        ctx.insert("operator_pass".into(), op.and_then(|c| c.password.clone()).unwrap_or_default());
        ctx.insert("operator_hash".into(), op.and_then(|c| c.nt_hash.clone()).unwrap_or_default());

        if let Some(f) = finding {
            if let Some(v) = f.get("target").and_then(|v| v.as_str()) {
                ctx.insert("finding.target".into(), v.to_string());
            }
            if let Some(v) = f.get("attack").and_then(|v| v.as_str()) {
                ctx.insert("finding.attack".into(), v.to_string());
            }
            if let Some(v) = f.get("severity").and_then(|v| v.as_str()) {
                ctx.insert("finding.severity".into(), v.to_string());
            }
            if let Some(data) = f.get("data").and_then(|v| v.as_object()) {
                for (k, v) in data {
                    ctx.insert(format!("finding.data.{k}"), value_to_template_string(v));
                }
            }
        }
        ctx
    }

    /// Write the loot + history to `run_dir/journal.json`. Matches the
    /// Python writer field-for-field; verified by the parity test.
    pub fn write_journal(&self) -> Result<PathBuf> {
        let path = self.run_dir.join("journal.json");

        #[derive(Serialize)]
        struct LootPayload<'a> {
            credentials:  &'a [Credential],
            tickets:      &'a [Ticket],
            certificates: &'a [Certificate],
            owned_hosts:  &'a [OwnedHost],
            files:        BTreeMap<String, String>,
        }
        #[derive(Serialize)]
        struct Payload<'a> {
            domain:     &'a str,
            dc_ip:      &'a str,
            #[serde(skip_serializing_if = "Option::is_none")]
            domain_sid: Option<&'a str>,
            loot:       LootPayload<'a>,
            history:    &'a [PlayRecord],
        }

        let files_str: BTreeMap<String, String> = self
            .loot
            .files
            .iter()
            .map(|(k, v)| (k.clone(), v.display().to_string()))
            .collect();

        let payload = Payload {
            domain: &self.domain,
            dc_ip:  &self.dc_ip,
            domain_sid: self.domain_sid.as_deref(),
            loot: LootPayload {
                credentials:  &self.loot.credentials,
                tickets:      &self.loot.tickets,
                certificates: &self.loot.certificates,
                owned_hosts:  &self.loot.owned_hosts,
                files:        files_str,
            },
            history: &self.history,
        };
        let serialized = serde_json::to_string_pretty(&payload)?;
        std::fs::write(&path, serialized)?;
        Ok(path)
    }
}


// ─────────────────────────────────────────────────────────────────── //
//  Helpers                                                            //
// ─────────────────────────────────────────────────────────────────── //


fn now_iso() -> String {
    let now: DateTime<Utc> = Utc::now();
    now.format("%Y-%m-%dT%H:%M:%S+00:00").to_string()
}


fn default_run_dir(domain: &str) -> PathBuf {
    let ts = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let domain_part = if domain.is_empty() {
        "engagement".to_string()
    } else {
        domain.replace('.', "_")
    };
    let base: PathBuf = std::env::var_os("KERB_CHAIN_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            std::env::var_os("HOME")
                .map(PathBuf::from)
                .unwrap_or_else(|| PathBuf::from("."))
                .join(".kerb-chain")
        });
    base.join("runs").join(format!("{domain_part}_{ts}"))
}


fn value_to_template_string(v: &serde_json::Value) -> String {
    match v {
        serde_json::Value::String(s) => s.clone(),
        serde_json::Value::Null      => String::new(),
        other                        => other.to_string(),
    }
}


// ─────────────────────────────────────────────────────────────────── //
//  Path helper visible to runner                                      //
// ─────────────────────────────────────────────────────────────────── //


/// Resolve a (possibly templated, possibly relative) path against the
/// engagement's `run_dir`. Used by capture rules for file output.
pub fn resolve_under_run_dir(run_dir: &Path, candidate: impl AsRef<Path>) -> PathBuf {
    let p = candidate.as_ref();
    if p.is_absolute() {
        p.to_path_buf()
    } else {
        run_dir.join(p)
    }
}
